### PR TITLE
Add spend-index feature for per-outpoint transparent spend queries

### PR DESCRIPTION
**Pull request by nuttycom**
Adds a `spend-index` cargo feature (in zcash_client_backend and zcash_client_sqlite) for wallets whose chain-data source can resolve the spend of an individual transparent outpoint directly (e.g. a full node with a spent-outpoint index), as an alternative to address-based `TransactionsInvolvingAddress` requests.

- zcash_client_backend: gated `TransactionDataRequest::SpendsOfTransparentOutput(OutPoint)` request variant and `WalletWrite::notify_outpoint_unspent_checked`. The servicing contract documents that spentness should be taken from an authoritative source and the spend index used only to resolve the spender, with "spent but spender not yet resolvable" treated as retryable (see ZcashFoundation/zebra#10806).
- zcash_client_sqlite: when `spend-index` is enabled, `transaction_data_requests` emits the per-outpoint variant for spend detection instead of `TransactionsInvolvingAddress` (retained for ephemeral-address discovery and when the feature is disabled); implements `notify_outpoint_unspent_checked` by advancing the per-outpoint `max_observed_unspent_height` watermark.
- zcash_transparent: derive `Hash` for `OutPoint` (required for the `TransactionDataRequest` `Hash` derive; also a natural map/set key).

A unit test for the new emission/notify path is left as a TODO (noted in transparent.rs): populating the spend-search queue requires the low-level/block-scan store path.

### CODE REVIEW

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 1]
```diff
@@ -1648,6 +1650,47 @@ pub(crate) fn update_observed_unspent_heights<P: consensus::Parameters>(
...
+            ELSE max_observed_unspent_height
+         END
+         WHERE tro.output_index = :output_index
+         AND tro.transaction_id IN (
+             SELECT id_tx FROM transactions WHERE txid = :txid
+         )
+         AND tro.id NOT IN (
+             SELECT transparent_received_output_id
+             FROM transparent_received_output_spends
+         )",
```
Comment (nuttycom) [inline zcash_client_sqlite/src/wallet/transparent.rs]: Could this be implemented more efficiently using the `UPDATE...FROM` sqlite3 extension?

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 2]
```diff
@@ -1820,58 +1863,101 @@ pub(crate) fn transaction_data_requests<P: consensus::Parameters>(
...
+            "SELECT t.txid, ssq.output_index
+             FROM transparent_spend_search_queue ssq
+             JOIN transactions t ON t.id_tx = ssq.transaction_id
+             JOIN transparent_received_outputs tro
+                ON tro.transaction_id = ssq.transaction_id AND tro.output_index = ssq.output_index
+             JOIN addresses ON addresses.id = tro.address_id
+             LEFT OUTER JOIN transparent_received_output_spends tros
+                ON tros.transparent_received_output_id = tro.id
+             WHERE tros.transaction_id IS NULL
+             AND addresses.key_scope != :ephemeral_key_scope
```
Comment (nuttycom) [inline zcash_client_sqlite/src/wallet/transparent.rs]: Since we're assuming privacy-preserving retrieval of spend information, I don't think we need to exclude ephemeral address-associated keys here.

File: zcash_client_backend/src/data_api.rs [hunk 3]
```diff
@@ -1106,10 +1106,14 @@ pub enum TransactionDataRequest {
...
     ///
     /// Fully transparent transactions, and transactions that do not contain either shielded inputs
-    /// or shielded outputs belonging to the wallet, may not be discovered by the process of chain
-    /// scanning; as a consequence, the wallet must actively query to find transactions that spend
-    /// such funds. Ideally we'd be able to query by [`OutPoint`] but this is not currently
-    /// functionality that is supported by the light wallet server.
+    /// or shielded outputs belonging to the wallet, may not be discovered by the process of
+    /// out-of-order chain scanning due to race conditions related to advancing the transparent
+    /// address gap limit; as a consequence, the wallet must actively query to find transactions
+    /// that spend such funds. Ideally we'd be able to query by [`OutPoint`] but this is not
```
Comment (nullcopy) [inline zcash_client_backend/src/data_api.rs:1112]: This note about what the lightwallet server currently supports feels unnecessary here. This would be a useful comment at the 'spend-index' feature: e.g. "only enable this feature if the lightwallet server supports it"

File: zcash_client_backend/src/data_api.rs [hunk 4]
```diff
@@ -1120,6 +1124,29 @@ pub enum TransactionDataRequest {
...
+    /// authoritative source (e.g. the node's UTXO set); the spend index is used only to identify
+    /// the spending transaction. If the output is spent, the caller should provide the spending
+    /// transaction to [`wallet::decrypt_and_store_transaction`]. If the output is confirmed
+    /// unspent as of some height, the caller should invoke
+    /// [`WalletWrite::notify_output_verified_unspent`] with that height. If the output is known to
+    /// be spent but the spending transaction cannot yet be resolved (e.g. an index is still being
+    /// built — see ZcashFoundation/zebra#10806), the caller should do nothing, so that the request
+    /// is re-issued and retried later.
+    #[cfg(feature = "spend-index")]
+    GetSpendingTx(OutPoint),
```
Comment (nullcopy) [inline zcash_client_backend/src/data_api.rs:1149]: Might be nice to make this a batch query

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 5]
```diff
@@ -1648,6 +1650,47 @@ pub(crate) fn update_observed_unspent_heights<P: consensus::Parameters>(
...
+/// Sets the max observed unspent height for the unspent transparent output identified by the given
+/// outpoint to at least the given height (will not cause the height to decrease). Used to record
+/// the result of a [`TransactionDataRequest::GetSpendingTx`] check that found the
+/// output unspent.
+///
+/// [`TransactionDataRequest::GetSpendingTx`]: zcash_client_backend::data_api::TransactionDataRequest::GetSpendingTx
+#[cfg(feature = "spend-index")]
+pub(crate) fn update_observed_unspent_height_for_outpoint(
+    conn: &rusqlite::Transaction,
+    outpoint: &OutPoint,
```
Comment (nullcopy) [inline zcash_client_sqlite/src/wallet/transparent.rs:1662]: Same comment as above, might be nice to support batch queries

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 6]
```diff
@@ -1648,6 +1650,47 @@ pub(crate) fn update_observed_unspent_heights<P: consensus::Parameters>(
...
+    checked_at: BlockHeight,
+) -> Result<(), SqliteClientError> {
+    let chain_tip_height = chain_tip_height(conn)?.ok_or(SqliteClientError::ChainHeightUnknown)?;
+    let checked_at = std::cmp::min(checked_at, chain_tip_height);
```
Comment (nullcopy) [inline zcash_client_sqlite/src/wallet/transparent.rs:1666]: Why do we care if the checked height is above the chain tip? This feels like it could easily trip if we race a transaction that was confirmed in the latest block, before we may have updated our tip.

This will force us to retry after the tip updates, which doesn't seem necessary, since the lightwallet server is trusted and providing the tip updates as well

File: zcash_client_backend/src/data_api.rs [hunk 7]
```diff
@@ -1120,6 +1124,29 @@ pub enum TransactionDataRequest {
...
     #[cfg(feature = "transparent-inputs")]
     TransactionsInvolvingAddress(TransactionsInvolvingAddress),
+    /// The main-chain transaction that spends of a specific transparent output, or confirmation
+    /// that the output remains unspent, is requested.
+    ///
+    /// When the `spend-index` feature is enabled, `GetSpendingTx` requests will be emitted by the
+    /// backend instead of [`TransactionDataRequest::TransactionsInvolvingAddress`], in
+    /// circumstances when spentness determination is needed. This feature should only be enabled
+    /// for wallets whose chain-data source can resolve the spend of an individual outpoint
+    /// directly — for example a full node that maintains a spent-outpoint index. It avoids needing
```
Comment (dannywillems) [inline zcash_client_backend/src/data_api.rs:1134]: > for example a full node that maintains a spent-outpoint index

I guess we expect that in Zaino? I see this comment [here](https://github.com/zcash/zallet/issues/507)

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 6 above)
Comment (dannywillems) [inline zcash_client_sqlite/src/wallet/transparent.rs:1666]: Comment: this handles reorg.

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 8]
```diff
@@ -1648,6 +1650,47 @@ pub(crate) fn update_observed_unspent_heights<P: consensus::Parameters>(
...
+
+    let mut stmt = conn.prepare(
+        "UPDATE transparent_received_outputs AS tro
+         SET max_observed_unspent_height = CASE
```
Comment (nullcopy) [inline zcash_client_sqlite/src/wallet/transparent.rs:1670]: Instead of casing the check-height, would it be better to filter `WHERE max_observed_unspent_height IS NULL OR max_observed_unspent_height < :checked_at`? The same criteria you use below to decide which outpoints to query

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 6 above)
Comment (dannywillems) [reply zcash_client_sqlite/src/wallet/transparent.rs:1666]: I guess it is expected.

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 9]
```diff
@@ -1820,58 +1863,101 @@ pub(crate) fn transaction_data_requests<P: consensus::Parameters>(
...
+    // Create transaction data requests that can find spends of our received UTXOs.
+    //
+    // With the `spend-index` feature, the chain-data source can resolve the spend of an
+    // individual outpoint directly, so we request spends per-outpoint. Otherwise we fall back to
+    // address-based requests (which, so long as address-based transaction data requests are
+    // required at all, are served by address-based lookups rather than by querying the spends of
+    // the associated outpoints directly).
+    #[cfg(feature = "spend-index")]
+    let spend_search_requests = {
+        // Per-outpoint spend resolution is privacy-preserving (it does not correlate the
```
Comment (nullcopy) [inline zcash_client_sqlite/src/wallet/transparent.rs:1875]: > Per-outpoint spend resolution is privacy-preserving (it does not correlate the wallet's addresses to an untrusted server), so unlike the address-based path below there is no need to exclude ephemeral-address outpoints here.

This statement is only true if the requests are not temporally correlated and there is sufficient cover traffic at to the lightwallet server. I would not personally be comfortable with having my ephemeral outpoints queried in the same process.

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 10]
```diff
@@ -2428,6 +2516,39 @@ mod tests {
...
+        // Exercises the `spend-index` `UPDATE ... FROM` in
+        // `update_observed_unspent_height_for_outpoint` (the outpoint matches no rows).
+        let tx = st.wallet().db().conn.unchecked_transaction().unwrap();
+        super::update_observed_unspent_height_for_outpoint(
+            &tx,
+            &OutPoint::new([1u8; 32], 0),
+            chain_tip,
+        )
+        .unwrap();
+        tx.commit().unwrap();
```
Comment (nullcopy) [inline zcash_client_sqlite/src/wallet/transparent.rs:2549]: Shouldn't you check the spend height was updated?

File: zcash_client_backend/src/data_api.rs (hunk 7 above)
Comment (nuttycom) [reply zcash_client_backend/src/data_api.rs:1134]: It's available in Zebra now with the `-indexer` feature enabled. There are some problems with it, but it's available.

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 10 above)
Comment (nullcopy) [reply zcash_client_sqlite/src/wallet/transparent.rs:2549]: Also need to test the negative case; attempting to reduce the spend height should be noop

**Review by nullcopy — CHANGES_REQUESTED** [review]
Blocking 4cdba554f5545050da784fc9e1eb073ae958d819 to discuss privacy implications

Other comments are nice-to-have (IMO), but not blocking

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 6 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/wallet/transparent.rs:1666]: I think it's better to have to re-check than it is to reference a block height that we don't know about anywhere; I don't know what invariants referencing such a height would break and I don't want to have to think about it. Remember that the "mempool height" or "target height" is used in a number of places, and we don't want to be in a situation where target heights could be confused for actual block heights.

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 9 above)
Comment (nuttycom) [reply zcash_client_sqlite/src/wallet/transparent.rs:1875]: The whole gate for the spend-index feature is "I can retrieve spentness information in a privacy-preserving manner".

File: zcash_client_sqlite/src/wallet/transparent.rs [hunk 11]
```diff
@@ -1820,58 +1863,101 @@ pub(crate) fn transaction_data_requests<P: consensus::Parameters>(
...
+    // required at all, are served by address-based lookups rather than by querying the spends of
+    // the associated outpoints directly).
+    #[cfg(feature = "spend-index")]
+    let spend_search_requests = {
+        // Per-outpoint spend resolution is privacy-preserving (it does not correlate the
+        // wallet's addresses to an untrusted server), so unlike the address-based path below
+        // there is no need to exclude ephemeral-address outpoints here.
+        let mut spend_requests_stmt = conn.prepare_cached(
+            "SELECT t.txid, ssq.output_index
+             FROM transparent_spend_search_queue ssq
```
Comment (dannywillems) [inline zcash_client_sqlite/src/wallet/transparent.rs:1880]: Note/comment for me: this table is fed by `put_transaction_outputs`, with `queue_transaction_spend_detection`.

**Review by dannywillems — APPROVED** [review]
The SQL requests LGTM. I went through the code to check the different tables and how they were populated. They're behind `transparent-inputs`, which is activated by `spend-index`.

File: zcash_client_sqlite/src/wallet/transparent.rs (hunk 9 above)
Comment (nullcopy) [reply zcash_client_sqlite/src/wallet/transparent.rs:1875]: Ah, that was not clear to me. In that case, I have no further objections

**Review by nullcopy — APPROVED** [review]
@nuttycom and I discussed the requested changes, and agreed to defer/skip each
utACK 4cdba554f5545050da784fc9e1eb073ae958d819
